### PR TITLE
bump version to 0.6.14

### PR DESCRIPTION
**Pull request by aleozlx**
## Description

Bump version to 0.6.14 for release.

## Related Issues (Gated-by PRs)

https://github.com/flashinfer-ai/flashinfer/issues?q=is%3Aopen+label%3Av0.6.14

+ #3783

## Reviewer Notes

**API changes review**

API changes since v0.6.13, using `scripts/list_apis.sh`

```diff
diff -u \
  <(scripts/list_apis.sh -d -p --ref v0.6.13) \
  <(scripts/list_apis.sh -d -p)
--- /dev/fd/63	2026-06-29 05:10:19.631000026 +0000
+++ /dev/fd/62	2026-06-29 05:10:19.632000026 +0000
@@ -339,6 +339,10 @@
     top_k: int,
     combine_payload_offset: int,
     payload_in_workspace: bool = False,
+    output_dtype: Optional[torch.dtype] = None,
+    output_scales: Optional[torch.Tensor] = None,
+    output_scalar_scale: float = 1.0,
+    sf_layout: SfLayout = SfLayout.layout_linear,
 ) -> torch.Tensor:
 
 
@@ -385,6 +389,10 @@
         payload: torch.Tensor,
         runtime_max_tokens_per_rank: int,
         payload_in_workspace: bool = False,
+        output_dtype: Optional[torch.dtype] = None,
+        output_scales: Optional[torch.Tensor] = None,
+        output_scalar_scale: float = 1.0,
+        sf_layout: SfLayout = SfLayout.layout_linear,
     ) -> torch.Tensor:
 
     @flashinfer_api
@@ -1018,6 +1026,7 @@
     gemm1_alpha: Optional[torch.Tensor] = None,
     gemm1_beta: Optional[torch.Tensor] = None,
     gemm1_clamp_limit: Optional[torch.Tensor] = None,
+    output: Optional[torch.Tensor] = None,
 ) -> Union[List[torch.Tensor], torch.Tensor]:
 
 
@@ -1047,6 +1056,7 @@
     gemm1_alpha: Optional[torch.Tensor] = None,
     gemm1_beta: Optional[torch.Tensor] = None,
     gemm1_clamp_limit: Optional[torch.Tensor] = None,
+    output: Optional[torch.Tensor] = None,
 ) -> Union[torch.Tensor, List[torch.Tensor]]:
 
 
@@ -1076,6 +1086,7 @@
     activation_type: int = ActivationType.Swiglu.value,
     norm_topk_prob: bool = True,
     routing_replay_out: Optional[torch.Tensor] = None,
+    output: Optional[torch.Tensor] = None,
 ) -> Union[List[torch.Tensor], torch.Tensor]:
 
 
@@ -1110,6 +1121,7 @@
     gemm1_alpha: Optional[torch.Tensor] = None,
     gemm1_beta: Optional[torch.Tensor] = None,
     gemm1_clamp_limit: Optional[torch.Tensor] = None,
+    output: Optional[torch.Tensor] = None,
 ) -> Union[List[torch.Tensor], torch.Tensor]:
 
 
@@ -1345,7 +1357,7 @@
         hidden_size: int,
         intermediate_size: int,
         use_cuda_graph: bool = False,
-        max_num_tokens: int = 4096,
+        max_num_tokens: Optional[int] = None,
         num_local_experts: Optional[int] = None,
         local_expert_offset: int = 0,
         tile_size: int = 128,
@@ -1353,6 +1365,7 @@
         output_dtype: torch.dtype = torch.bfloat16,
         device: str = "cuda",
         enable_pdl: bool = True,
+        activation: str = "silu",
     ):
 
     @flashinfer_api(trace=cute_dsl_moe_wrapper_run_trace)
@@ -1395,6 +1408,7 @@
     moe_output: Optional[torch.Tensor] = None,
     aux_stream: Optional[torch.cuda.Stream] = None,
     enable_pdl: bool = True,
+    activation: str = "silu",
 ) -> torch.Tensor:
 [Global Functions]
 @flashinfer_api(trace=fused_topk_deepseek_trace)
@@ -1460,8 +1474,10 @@
     scale: Optional[float] = None,
     output: Optional[torch.Tensor] = None,
     intermediate_states_buffer: Optional[torch.Tensor] = None,
+    ssm_state_indices: Optional[torch.Tensor] = None,
     disable_state_update: Optional[bool] = None,
     use_qk_l2norm: bool = True,
+    output_state_indices: Optional[torch.Tensor] = None,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
 [Global Functions]
 @flashinfer_api(trace=gdn_prefill_trace)
@@ -1481,6 +1497,7 @@
     state_checkpoints: Optional[torch.Tensor] = None,
     checkpoint_cu_starts: Optional[torch.Tensor] = None,
     checkpoint_every_n_tokens: int = 0,
+    use_cp: Literal["auto"] | bool = "auto",
 ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
 [Global Functions]
 @flashinfer_api(trace=mm_bf16_trace)
@@ -1492,7 +1509,7 @@
     out: Optional[torch.Tensor] = None,
     out_dtype: torch.dtype = torch.bfloat16,
     backend: Literal[
-        "cudnn", "cutlass", "tgv", "cublaslt", "tinygemm", "auto"
+        "cudnn", "cutlass", "tgv", "cublaslt", "tinygemm", "cutile", "auto"
     ] = "cudnn",
 ) -> torch.Tensor:
     A: torch.Tensor,
@@ -1507,6 +1524,12 @@
     out_dtype: torch.dtype = torch.bfloat16,
     backend: Literal["cudnn", "cutlass", "auto"] = "cudnn",
 ):
+    A: torch.Tensor,
+    B: torch.Tensor,
+    out: Optional[torch.Tensor] = None,
+    out_dtype: torch.dtype = torch.bfloat16,
+    backend: Literal["cudnn", "cutlass", "cutile", "auto"] = "cudnn",
+):
 
 
     A: torch.Tensor,
@@ -1530,7 +1553,7 @@
     B: torch.Tensor,
     out: Optional[torch.Tensor] = None,
     out_dtype: torch.dtype = torch.bfloat16,
-    backend: Literal["cudnn", "cutlass", "auto"] = "cudnn",
+    backend: Literal["cudnn", "cutlass", "cutile", "auto"] = "cudnn",
 ) -> torch.Tensor:
 
 @flashinfer_api(trace=tgv_gemm_sm100_trace)
@@ -1703,6 +1726,17 @@
     out_dtype: Optional[torch.dtype] = None,
     backend: Literal["cutlass", "trtllm"] = "cutlass",
 ):
+    a: torch.Tensor,
+    b: torch.Tensor,
+    a_scale: torch.Tensor,
+    b_scale: torch.Tensor,
+    scale_major_mode: Optional[Literal["MN", "K"]] = None,
+    mma_sm: int = 1,
+    scale_granularity_mnk: Tuple[int, int, int] = (1, 128, 128),
+    out: Optional[torch.Tensor] = None,
+    out_dtype: Optional[torch.dtype] = None,
+    backend: Literal["cutlass", "trtllm", "cutile"] = "cutlass",
+):
 @flashinfer_api(trace=gemm_fp8_nt_groupwise_trace)
 def gemm_fp8_nt_groupwise(
     a: torch.Tensor,
@@ -1714,7 +1748,7 @@
     scale_granularity_mnk: Tuple[int, int, int] = (1, 128, 128),
     out: Optional[torch.Tensor] = None,
     out_dtype: Optional[torch.dtype] = None,
-    backend: Literal["cutlass", "trtllm"] = "cutlass",
+    backend: Literal["cutlass", "trtllm", "cutile"] = "cutlass",
 ) -> torch.Tensor:
 @flashinfer_api
 def gemm_fp8_nt_blockscaled(
@@ -1737,6 +1771,7 @@
     mma_sm: int = 1,
     out: Optional[torch.Tensor] = None,
     out_dtype: Optional[torch.dtype] = None,
+    backend: Literal["trtllm", "cutile"] = "trtllm",
 ):
 @flashinfer_api
 def group_gemm_fp8_nt_groupwise(
@@ -1750,6 +1785,7 @@
     mma_sm: int = 1,
     out: Optional[torch.Tensor] = None,  # (cum_m, n)
     out_dtype: Optional[torch.dtype] = None,
+    backend: Literal["trtllm", "cutile"] = "trtllm",
 ) -> torch.Tensor:
     a: torch.Tensor,
     b: torch.Tensor,
@@ -1880,6 +1916,31 @@
     backend: Literal["cudnn", "cutlass", "auto"] = "auto",
 ) -> torch.Tensor:
 [Global Functions]
+@flashinfer_api
+def prepare_bf16_fp4_weights(
+    b: torch.Tensor,
+    b_descale: torch.Tensor,
+    alpha: Optional[torch.Tensor] = None,
+    *,
+    backend: Literal["cudnn", "cute-dsl"],
+    block_size: int = 16,
+) -> Tuple[torch.Tensor, torch.Tensor, Optional[torch.Tensor]]:
+@flashinfer_api(trace=mm_bf16_fp4_trace_dispatch)
+def mm_bf16_fp4(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    b_descale: torch.Tensor,
+    alpha: Optional[torch.Tensor] = None,
+    *,
+    backend: Literal["cudnn", "cute-dsl"],
+    out_dtype: Optional[torch.dtype] = None,
+    out: Optional[torch.Tensor] = None,
+    block_size: int = 16,
+    enable_pdl: bool = True,
+) -> torch.Tensor:
+
+
+[Global Functions]
 @flashinfer_api(trace=grouped_gemm_nt_masked_trace)
 def grouped_gemm_nt_masked(
     lhs: Tuple[torch.Tensor, torch.Tensor],
@@ -2022,6 +2083,20 @@
 ) -> torch.Tensor:
 [Global Functions]
 @flashinfer_api
+def moe_gemm_mxfp8_nt_groupwise(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    a_scale: torch.Tensor,
+    b_scale: torch.Tensor,
+    m_indptr: torch.Tensor,
+    scale_granularity_mnk: Tuple[int, int, int] = (1, 1, 128),
+    scale_major_mode: Literal["MN"] = "MN",
+    backend: Literal["cute"] = "cute",
+    out: Optional[torch.Tensor] = None,
+    out_dtype: Optional[torch.dtype] = None,
+) -> torch.Tensor:
+[Global Functions]
+@flashinfer_api
 def recurrent_kda(
     q: torch.Tensor,
     k: torch.Tensor,
@@ -2335,11 +2410,16 @@
         page_table: Optional[torch.Tensor] = None,
         return_lse_base_on_e: bool = False,
         o_scale: Optional[float] = None,
+        *,
+        ckv_scale: Optional[float] = None,
+        kpe_scale: Optional[float] = None,
     ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
 
 
 
 
+
+
     workspace_bytes: int,
     q_len: int,
     num_heads: int,
@@ -2396,7 +2476,7 @@
     kv_lora_rank: int,
     qk_rope_head_dim: int,
     block_tables: torch.Tensor,
-    seq_lens: torch.Tensor,
+    seq_lens: Optional[torch.Tensor],
     max_seq_len: int,
     sparse_mla_top_k: int = 0,
     out: Optional[torch.Tensor] = None,
@@ -2411,6 +2491,9 @@
     lse: Optional[torch.Tensor] = None,
     return_lse: bool = False,
     cute_dsl_impl: str = "auto",
+    kv_scale_format: str = "auto",
+    cum_seq_lens_q: Optional[torch.Tensor] = None,
+    max_q_len: Optional[int] = None,
 ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
 
 
@@ -2711,6 +2794,50 @@
     v_out: Optional[torch.Tensor] = None,
 ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
 [Global Functions]
+@flashinfer_api(trace=nvfp4_attention_sm120_quantize_qkv_trace)
+def nvfp4_attention_sm120_quantize_qkv(
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    per_block_mean: bool = True,
+) -> Tuple[
+    torch.Tensor,
+    torch.Tensor,
+    torch.Tensor,
+    torch.Tensor,
+    torch.Tensor,
+    torch.Tensor,
+    torch.Tensor,
+]:
+
+
+    q_fp4: torch.Tensor,
+    k_fp4: torch.Tensor,
+    v_fp4_t: torch.Tensor,
+    q_scale: torch.Tensor,
+    k_scale: torch.Tensor,
+    v_scale_t: torch.Tensor,
+    qk_correction: torch.Tensor,
+    per_block_mean: bool,
+) -> Tuple[int, int, int, int]:
+@flashinfer_api(trace=nvfp4_attention_sm120_fwd_trace)
+def nvfp4_attention_sm120_fwd(
+    q_fp4: torch.Tensor,
+    k_fp4: torch.Tensor,
+    v_fp4_t: torch.Tensor,
+    q_scale: torch.Tensor,
+    k_scale: torch.Tensor,
+    v_scale_t: torch.Tensor,
+    qk_correction: torch.Tensor,
+    sm_scale: Optional[float] = None,
+    causal: bool = False,
+    per_block_mean: bool = True,
+    out: Optional[torch.Tensor] = None,
+    lse: Optional[torch.Tensor] = None,
+    out_dtype: torch.dtype = torch.bfloat16,
+    softmax_scale: Optional[float] = None,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+[Global Functions]
 @flashinfer_api
 def get_batch_indices_positions(
     append_indptr: torch.Tensor,
@@ -3396,6 +3523,11 @@
     backend: Literal["cuda", "cute-dsl"] = "cuda",
     sf_swizzle_layout: Optional[SfLayout] = None,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
+@flashinfer_api(trace=mxfp8_grouped_quantize_trace)
+def mxfp8_grouped_quantize(
+    a: torch.Tensor,
+    mask: torch.Tensor,
+) -> Tuple[torch.Tensor, torch.Tensor]:
 
 
 @flashinfer_api
@@ -3722,6 +3854,25 @@
 ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
 
 
+    x: torch.Tensor,
+    top_k: Union[torch.Tensor, int],
+    indices: Optional[torch.Tensor],
+) -> bool:
+
+
+    x: torch.Tensor,
+    top_k: int,
+    top_p: Union[torch.Tensor, float],
+    *,
+    from_logits: bool,
+    deterministic: bool,
+    generator: Optional[torch.Generator],
+    check_nan: bool,
+    seed: Optional[Union[int, torch.Tensor]],
+    offset: Optional[Union[int, torch.Tensor]],
+    return_valid: bool = False,
+):
+
 @flashinfer_api(trace=top_k_top_p_sampling_from_logits_trace)
 def top_k_top_p_sampling_from_logits(
     logits: torch.Tensor,
```

## Summary by CodeRabbit

* **Chores**
  * Updated the app version from 0.6.13 to 0.6.14.

### CODE REVIEW

**Comment by aleozlx** [comment]
i wonder if this can use existing enum

---

+ #3080

https://github.com/flashinfer-ai/flashinfer/blame/eb79c537b22bc42786ea369827736cff2f615b74/flashinfer/fused_moe/cute_dsl/b12x_moe.py#L75
```
+        activation: str = "silu",
```

---

this naming could be simplified

https://github.com/flashinfer-ai/flashinfer/blame/eb79c537b22bc42786ea369827736cff2f615b74/flashinfer/grouped_mm/cute_sm120_mxfp8_groupwise/core.py#L96

+ #3562

```
def moe_gemm_mxfp8_nt_groupwise(
```

**Review by bkryu — APPROVED** [review]
